When user clone/pull/etc a project which doesn't exist, print a nice message

JuneKelly opened the pull request:
The old behaviour was to just show an equivalent of a 500 error.
This way, we catch the 404 from the API, and print an appropriate message
to the user.

## Screenshots

End result (updated): 

![image](https://user-images.githubusercontent.com/1540054/49939223-201b9300-fed4-11e8-85bb-e13b739f9615.png)


Child of https://github.com/overleaf/sharelatex/issues/1311

Commit:
When user clone/pull/etc a project which doesn't exist, print a nice message

The old behaviour was to just show an equivalent of a 500 error.
This way, we catch the 404 from the API, and print an appropriate message
to the user.

jdleesmiller commented on src/main/java/uk/ac/ic/wlgitbridge/snapshot/base/MissingRepositoryException.java:
```diff
@@ -10,43 +10,40 @@
+        "because the project does not exist, or git access is not enabled for the project.",
         "",
-        "If this problem persists, please contact us."
+        "If this problem persists, please contact us at at support@overleaf.com"
```
Double `at`. Maybe we should also keep the link to 342 for more information? That page could explain how to enable git access, for example.

Hmm. Also, https://www.overleaf.com/help/342 is the general Overleaf v2 FAQ. Possibly we should change that to point to 343, which is https://www.overleaf.com/learn/how-to/Working_Offline_in_Overleaf, which will probably have more git-specific content.

jpallen replied:
> If this problem persists

I'm not sure about this wording - the problem will definitely persist :) How about 'if this is unexpected'?

jdleesmiller commented on src/main/java/uk/ac/ic/wlgitbridge/snapshot/base/MissingRepositoryException.java:
```diff
@@ -10,43 +10,40 @@
     public static final List<String> GENERIC_REASON = Arrays.asList(
-        "This Overleaf project currently has no git access.",
+        "This Overleaf project currently has no git access, either",
+        "because the project does not exist, or git access is not enabled for the project.",
```
I think `, or because` would read better here.

jpallen replied:
I think we could remove 'or git access is not enabled', since that's a different error right? This should only be for the 404 case, not the 403 case?

jpallen replied:
In fact, could this message just be short and sweet if it's just the 404 case:

<img width="584" alt="screen shot 2018-12-13 at 10 28 19" src="https://user-images.githubusercontent.com/31305/49932789-daee6580-fec1-11e8-89ad-18d4c4d48f5c.png">

WDYT?

jpallen replied:
Oh, I see that this may be generic message for both cases though... hmmm, is that hard to split out?

JuneKelly replied:
We'd need to audit every way a `40*` can arise in this path, not hard, but not keen to spend time on it at the moment.

jpallen replied:
Ok 👍

jdleesmiller commented on src/main/java/uk/ac/ic/wlgitbridge/git/handler/WLRepositoryResolver.java:
```diff
@@ -97,6 +98,8 @@ public Repository open(
             cannot occur
             */
+        } catch (MissingRepositoryException e) {
+            throw new ServiceMayNotContinueException(String.join("\n", e.getDescriptionLines()), e);
```
It took me a while to get my head around this, but hopefully I have now done so.

- [`ServiceMayNotContinueException `](http://download.eclipse.org/jgit/docs/jgit-2.0.0.201206130900-r/apidocs/org/eclipse/jgit/transport/ServiceMayNotContinueException.html) is an internal JGit exception, and this method is implementing an internal JGit interface, so we can't throw anything else. The docs say that the message for `ServiceMayNotContinueException` is intended for the end user, so that's good (matches what we are doing).
- `MissingRepositoryException` is one of ours, and it derives from `GitUserException` (via `SnapshotAPIException`, though I am not entirely sure that is right, because it doesn't do anything in the `fromJSON` method it needs to have because of `SnapshotAPIException`).
- `GitUserException` is sent to the user as a message and then a series of lines (the description), as part of the receive-pack protocol, in particular via `handleSnapshotPostException` in `WritelatexPutHook`. (It also gets sent as a series of lines from the `Oauth2Filter`, this time without the message, I think because git adds a helpful message in this case anyway.)
- This try-catch block handles `GitUserException` below, but it only passes back the message, not the description lines. This seems a bit inconsistent; I don't see a particular reason not to send back a multi-line message in this case.

So... another option here would be to change the `GitUserException` handler so that it also sends back the lines. That would also potentially improve the error messages in some other cases. WDYT?

JuneKelly replied:
Yup, that's basically it. And a good suggestion :+1:

jdleesmiller commented on src/main/java/uk/ac/ic/wlgitbridge/snapshot/base/MissingRepositoryException.java:
```diff
@@ -10,43 +10,40 @@
 
     private List<String> descriptionLines;
 
-    public MissingRepositoryException() {
```
Another option would be to keep this constructor and just have it set `descriptionLines` to `GENERIC_REASON`.

jdleesmiller reviewed (changes requested):
Good start!

Commit:
Fix wording of the generic-error when a repository is missing

Commit:
Keep the default constructor for MissingRepositoryException

Commit:
Handle GitUserException, with multiple lines of output

ShaneKilkelly commented:
@jdleesmiller up for re-review :)

jdleesmiller commented on src/main/java/uk/ac/ic/wlgitbridge/git/handler/WLRepositoryResolver.java:
```diff
@@ -13,6 +13,7 @@
 import uk.ac.ic.wlgitbridge.server.GitBridgeServer;
 import uk.ac.ic.wlgitbridge.server.Oauth2Filter;
 import uk.ac.ic.wlgitbridge.snapshot.base.ForbiddenException;
+import uk.ac.ic.wlgitbridge.snapshot.base.MissingRepositoryException;
```
This should now be redundant?

jdleesmiller commented on src/main/java/uk/ac/ic/wlgitbridge/git/handler/WLRepositoryResolver.java:
```diff
@@ -109,7 +110,7 @@ public Repository open(
             throw new ServiceNotAuthorizedException();
         } catch (GitUserException e) {
-            throw new ServiceMayNotContinueException(e.getMessage(), e);
+            throw new ServiceMayNotContinueException(String.join("\n", e.getDescriptionLines()), e);
```
This completely omits the message, which may be informative. Maybe just add the message on its own line before the description?

jdleesmiller commented on src/main/java/uk/ac/ic/wlgitbridge/snapshot/base/MissingRepositoryException.java:
```diff
@@ -10,42 +10,42 @@
     public static final List<String> GENERIC_REASON = Arrays.asList(
-        "This Overleaf project currently has no git access.",
+        "This Overleaf project currently has no git access, either",
+        "because the project does not exist, or because git access is not enabled for the project.",
```
What is the target line length we're going for here? This one is 89. I think probably 72 characters?

JuneKelly replied:
Yeah, it's made a bit more tricky by the git client prefixing something like `remote error: something:` on the first line. Will try to balance them out

jdleesmiller replied:
Ah yes, good point. Agree on balancing a bit more, however.

jdleesmiller reviewed (changes requested):
Looks good. Minor comments.

Commit:
Remove stray import

Commit:
Include message along with error description lines

Commit:
Wrap the generic repository-missing message correctly

Now appears to the client as:

```
fatal: remote error: no git access
This Overleaf project currently has no git access, either because
the project does not exist, or because git access is not enabled
for the project.
```

ShaneKilkelly commented:
Updated view of what the client sees with a 404: 

![image](https://user-images.githubusercontent.com/1540054/49939223-201b9300-fed4-11e8-85bb-e13b739f9615.png)

ShaneKilkelly commented:
@jdleesmiller up for re-review :)

jdleesmiller reviewed (approved):
👍 thanks
